docs(listeners): Updates ingress and route listener procedures (#7842)
strimzi/strimzi-kafka-operator: documentation/modules/security/proc-accessing-kafka-using-ingress.adoc

// assembly-configuring-kafka-listeners.adoc

[id='proc-accessing-kafka-using-ingress-{context}']

> = Accessing Kafka using ingresses

Reviewer: TBH, I'm not sure this sounds right. The access is facilitated through a single Ingress controller or multiple Ingress resources. So maybe we can put there either `Accessing Kafka using Ingress controller` or `Accessing Kafka using ingress resources`?

Author: Right. Let's say _Accessing Kafka using an Nginx Ingress Controller_

Reviewer: `Accessing Kafka using a Kubernetes Nginx Ingress Controller` ;-)

[role="_abstract"]
Use an {NginxIngressController} to access a Strimzi Kafka cluster from clients outside the Kubernetes cluster. 

> The Nginx Ingress Controller supports connections from HTTP and HTTPS-based applications.

Reviewer: This is confusing. You should either remove it or elaborate on the TLS passthrough and its challenges right here. Because otherwise people might look at this and instead of continue reading say hey, I can use my HTTP client here.

Author: Removed

> To be able to use an Nginx Ingress Controller, add configuration for an `ingress` type listener in the `Kafka` custom resource. 

Reviewer: There are two Nginx Ingress controllers. The one we support is this one from Kubernetes: https://github.com/kubernetes/ingress-nginx ... AFAIK, it is not always the same as the Nginx Ingress Controller from NGINX company (https://docs.nginx.com/nginx-ingress-controller/). So we need to be clear about this in all places.

```suggestion
To be able to use a Kubernetes Nginx Ingress Controller, add configuration for an `ingress` type listener in the `Kafka` custom resource. 
```

When applied, the configuration creates a dedicated ingress and service for an external bootstrap and each broker in the cluster. 
Clients connect to the bootstrap ingress, which routes them through the bootstrap service to connect to a broker. 
Per-broker connections are then established using DNS names, which route traffic from the client to the broker through the broker-specific ingresses and services.

> To connect to a broker, you specify a hostname (advertised address) for the ingress bootstrap address, as well as the certificate used for authentication.

Reviewer: * The bootstraps are not advertised
* The mandatory part is the TLS encryption. Authentication is fully optional

```suggestion
To connect to a broker, you specify a hostname for the ingress bootstrap address, as well as the TLS certificate, used authenticate the Kafka brokers.
```

> For access using an ingress, the port is always 443.

Reviewer: I think we need to distinguish:
* The port in the Kafka CR will be something like 9094 for example. Not 443.
* Port 443 is typically used by the Ingress controller, so it will be used by the clients in their configuration (there might be situations when it is configured for other ports, but that is a more niche thing).

```suggestion
For access using an ingress, the port used in the Kafka client is typically 443.
```

.TLS passthrough

> Make sure that you enable TLS passthrough in your NGINX Ingress Controller deployment.

Reviewer: ```suggestion
Make sure that you enable TLS passthrough in your Kubernetes NGINX Ingress Controller deployment.
```

Author: I've given the full name in the initial reference (`Ingress NGINX Controller for Kubernetes`) with link and thereafter the shortname `Ingress NGINX Controller`.

Reviewer: TBH, I'm not sure this is enough.

Author: Okay. Let's go with the full name in the title and wherever it's mentioned. I don't know if we want to make it clear that it's only this controller that's supported?

Kafka uses a binary protocol over TCP, but the NGINX Ingress Controller is designed to work with a HTTP protocol. 

> To be able to route traffic through ingresses, Strimzi uses TLS passthrough with Server Name Indication (SNI).

Reviewer: ```suggestion
To be able to route TCP traffic through ingresses, Strimzi uses TLS passthrough with Server Name Indication (SNI).
```

SNI helps with identifying and passing connection to Kafka brokers.
In passthrough mode, TLS encryption is always used.

> Because connection passes to the brokers, the listeners use the TLS certificates signed by the internal cluster CA.

Reviewer: ```suggestion
Because the connection passes to the brokers, the listeners use the TLS certificates signed by the cluster CA and not the Ingress certificates.
```

To configure listeners to use your own listener certificates, xref:proc-installing-certs-per-listener-{context}[use the `brokerCertChainAndKey` property]. 

For more information about enabling TLS passthrough, see the {NginxIngressControllerTLSPassthrough}.

.Prerequisites

> * An NGINX Ingress Controller is running with TLS passthrough enabled

Reviewer: ```suggestion
* An Kubernetes NGINX Ingress Controller is running with TLS passthrough enabled
```

* A running Cluster Operator

In this procedure, the Kafka cluster name is `my-cluster`.

> NOTE: You can install and use the NGINX Ingress Operator to deploy an NGINX Ingress Controller.

Reviewer: I think this is the part which is NGINX only? I'm not aware of the Kubernetes Nginx Controller having an operator.

Author: Removed

            host: broker-1.myingress.com
          - broker: 2
            host: broker-2.myingress.com

>           class: nginx  

Reviewer: Should we add here a callout and clarify what it is and that it is optional if the user has set up a default ingress class?

+
`ClusterIP` type services are created for each Kafka broker, as well as an external bootstrap service.
+
An `ingress` is also created for each service, with a DNS address to expose them using the NGINX Ingress Controller.
+
.Ingresses created for the bootstrap and brokers
[source,shell]
----
NAME                        CLASS  HOSTS                    ADDRESS               PORTS
my-cluster-kafka-0          nginx  broker-0.myingress.com   external.ingress.com  80,443
my-cluster-kafka-1          nginx  broker-1.myingress.com   external.ingress.com  80,443
my-cluster-kafka-2          nginx  broker-2.myingress.com   external.ingress.com  80,443
my-cluster-kafka-bootstrap  nginx  bootstrap.myingress.com  external.ingress.com  80,443
----
+
The DNS addresses used for client connection are propagated to the `status` of each ingress.
+
.Status for the bootstrap ingress
[source,yaml]
----
status:
  loadBalancer:
    ingress:
      - hostname: external.ingress.com
 # ...
----

. Use a target broker to check the client-server TLS connection on port 443 using the OpenSSL `s_client`.  
+
[source,shell]
----

> openssl s_client -connect my-cluster-kafka-0:443 -servername my-cluster-kafka-0 -showcerts

Reviewer: Should this use the example addresses from above? E.g. `broker-0.myingress.com` instead of `my-cluster-kafka-0`?

----
+
The server name is the SNI for passing the connection to the broker. 
+
If the connection is successful, the certificates for the broker are returned.
+

> .Certificates for the broker

Author: This worked for routes, but was returning controller certs for ingress. I think it's useful. Assuming it would work in the same way here.

Reviewer: Yes, it should work the same way. When it doesn't it typically means that TLS passthrough is not enabled.
